### PR TITLE
[2/4] Intel GPU Runtime Upstreaming for Device

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #117619
* #117611
* #116869
* #116850
* __->__ #116833

# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), as mentioned in [[RFC] Intel GPU Runtime Upstreaming](https://github.com/pytorch/pytorch/issues/114842), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdxFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/xpu/detail/XPUHooks.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+bool XPUHooks::hasXPU() const {
+  return true;
+}
+
+std::string XPUHooks::showConfig() const {
+  return "XPU backend";
+}
+
+int XPUHooks::getGlobalIdxFromDevice(const at::Device& device) const {
```
Comment (EikanWang) [inline aten/src/ATen/xpu/detail/XPUHooks.cpp:20]:
```suggestion
DeviceIndex XPUHooks::getGlobalIdxFromDevice(const at::Device& device) const {
```

File: aten/src/ATen/xpu/XPUContext.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+void initXPUContextVectors() {
+  num_gpus = c10::xpu::device_count();
+  device_prop_flags.resize(num_gpus);
+  device_properties.resize(num_gpus);
+  device_global_index_flags.resize(num_gpus);
+  device_global_indexs.resize(num_gpus);
+}
+
+void initDeviceProperty(int device) {
```
Comment (EikanWang) [inline aten/src/ATen/xpu/XPUContext.cpp:36]:
```suggestion
void initDeviceProperty(DeviceIndex device) {
```

File: aten/src/ATen/xpu/XPUContext.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  DeviceProp device_prop;
+  c10::xpu::get_device_properties(&device_prop, device);
+  device_properties[device] = device_prop;
+}
+
+void initDeviceGlobalId(int device) {
```
Comment (EikanWang) [inline aten/src/ATen/xpu/XPUContext.cpp:42]:
```suggestion
void initDeviceGlobalIdx(DeviceIndex device) {
```

File: aten/src/ATen/xpu/XPUContext.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  sycl::device& raw_device = c10::xpu::get_raw_device(device);
+  // Get all SYCL devices associated with the SYCL platform.
+  auto devices = sycl::device::get_devices();
+  auto match_device = [raw_device](const auto& device) -> bool {
+    return raw_device == device;
+  };
+  auto it = std::find_if(devices.begin(), devices.end(), match_device);
+  TORCH_CHECK(it != devices.end(), "Cant't find the global id of XPU device.");
```
Comment (EikanWang) [inline aten/src/ATen/xpu/XPUContext.cpp:50]:
```suggestion
  TORCH_CHECK(it != devices.end(), "Cant't find the global index of XPU device.");
```

File: aten/src/ATen/xpu/XPUContext.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+} // anonymous namespace
+
+DeviceProp* getCurrentDeviceProperties() {
+  auto device = c10::xpu::current_device();
+  return getDeviceProperties(device);
+}
+
+DeviceProp* getDeviceProperties(int device) {
```
Comment (EikanWang) [inline aten/src/ATen/xpu/XPUContext.cpp:62]:
```suggestion
DeviceProp* getDeviceProperties(DeviceIndex device) {
```

File: aten/src/ATen/xpu/XPUContext.cpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      "device is out of range, device is ",
+      device,
+      ", total number of device is ",
+      num_gpus,
+      ".");
+  c10::call_once(device_prop_flags[device], initDeviceProperty, device);
+  return &device_properties[device];
+}
+
+int getGlobalIdxFromDevice(int device) {
```
Comment (EikanWang) [inline aten/src/ATen/xpu/XPUContext.cpp:77]:
```suggestion
DeviceIndex getGlobalIdxFromDevice(DeviceIndex device) {
```

File: aten/src/ATen/xpu/detail/XPUHooks.h [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace at::xpu::detail {
+
+// The real implementation of XPUHooksInterface
+struct XPUHooks : public at::XPUHooksInterface {
+  XPUHooks(at::XPUHooksArgs) {}
+  void initXPU() const override;
+  bool hasXPU() const override;
+  std::string showConfig() const override;
+  int getGlobalIdxFromDevice(const at::Device& device) const override;
```
Comment (EikanWang) [inline aten/src/ATen/xpu/detail/XPUHooks.h:16]:
```suggestion
  DeviceIndex getGlobalIdxFromDevice(const at::Device& device) const override;
```

File: aten/src/ATen/xpu/XPUContext.h [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+// XPU is available if we compiled with XPU.
+inline bool is_available() {
+  return c10::xpu::device_count() > 0;
+}
+
+TORCH_XPU_API DeviceProp* getCurrentDeviceProperties();
+
+TORCH_XPU_API DeviceProp* getDeviceProperties(int device);
+
+TORCH_XPU_API int getGlobalIdxFromDevice(int device);
```
Comment (EikanWang) [inline aten/src/ATen/xpu/XPUContext.h:16]: The two APIs are global, right? Then, should we use INT to align with other backends? Or should we use unsigned or `DeviceIndex`?

Comment (guangyey) [reply aten/src/ATen/xpu/XPUContext.h:16]: Align with other backends. use `int` in public APIs.

File: aten/src/ATen/xpu/XPUContext.cpp (hunk 4 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUContext.cpp]: done.

File: aten/src/ATen/detail/XPUHooksInterface.h [hunk 9]
```diff
@@ -9,11 +9,6 @@
 #include <functional>
 #include <memory>
 
-// We use forward declaration here instead of #include <ATen/dlpack.h> to avoid
-// leaking DLPack implementation detail to every project that includes `ATen/Context.h`, which in turn
-// would lead to a conflict when linked with another project using DLPack (for example TVM)
-struct DLDevice_;
-
```
Comment (EikanWang) [inline aten/src/ATen/detail/XPUHooksInterface.h:16]: These changes should not be part of this PR.

File: aten/src/ATen/detail/XPUHooksInterface.h [hunk 10]
```diff
@@ -44,23 +39,8 @@ struct TORCH_API XPUHooksInterface {
...
-  }
-
-  virtual DLDevice_& getDLPackDeviceFromATenDevice(
-      DLDevice_& dl_device,
-      const Device& aten_device,
-      void* data) const {
-    TORCH_CHECK(
-        false,
-        "Cannot get XPU DL device without Intel Extension for Pytorch. ",
-        XPU_HELP);
```
Comment (EikanWang) [inline aten/src/ATen/detail/XPUHooksInterface.h:63]: Ditto

**Review by EikanWang — CHANGES_REQUESTED** [review]

File: aten/src/ATen/xpu/XPUContext.cpp [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+  device_global_idxs.resize(num_gpus);
+}
+
+void initDeviceProperty(int device) {
+  DeviceProp device_prop;
+  c10::xpu::get_device_properties(&device_prop, device);
+  device_properties[device] = device_prop;
+}
+
+void initDeviceGlobalIdxs(int device) {
```
Comment (gujinghui) [inline aten/src/ATen/xpu/XPUContext.cpp:42]: The plural of index should be indexes? initDeviceGlobalIdx is good enough.

File: aten/src/ATen/xpu/detail/XPUHooks.cpp (hunk 1 above)
Comment (guangyey) [reply aten/src/ATen/xpu/detail/XPUHooks.cpp:20]: `device_id` in dlpack is `int` data type.

File: aten/src/ATen/xpu/XPUContext.cpp (hunk 2 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUContext.cpp]: done.

File: aten/src/ATen/xpu/XPUContext.cpp (hunk 5 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUContext.cpp]: done.

File: aten/src/ATen/xpu/XPUContext.cpp (hunk 6 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUContext.cpp]: done.

File: aten/src/ATen/xpu/detail/XPUHooks.h (hunk 7 above)
Comment (guangyey) [reply aten/src/ATen/xpu/detail/XPUHooks.h]: done.

File: aten/src/ATen/xpu/XPUContext.cpp (hunk 3 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUContext.cpp]: done.

File: aten/src/ATen/DLConvertor.cpp [hunk 12]
```diff
@@ -101,8 +101,9 @@ static DLDevice getDLDevice(const Tensor& tensor, const int64_t& device_id) {
       ctx.device_type = DLDeviceType::kDLROCM;
       break;
     case DeviceType::XPU:
-      ctx = at::detail::getXPUHooks().getDLPackDeviceFromATenDevice(
-          ctx, tensor.device(), tensor.data_ptr());
+      ctx.device_type = DLDeviceType::kDLOneAPI;
+      ctx.device_id =
+          at::detail::getXPUHooks().getGlobalIdxFromDevice(tensor.device());
```
Comment (guangyey) [inline aten/src/ATen/DLConvertor.cpp:106]: `kDLOneAPI` is a device type from DLPack, XPU uses it here. To avoid leaking DLPack implementation detail to every project that includes `ATen/Context.h`, we decouple `DLDevice_` from `XPUHooks` and only use the index of `DLDevice_` as our input.

File: aten/src/ATen/detail/XPUHooksInterface.h (hunk 9 above)
Comment (guangyey) [reply aten/src/ATen/detail/XPUHooksInterface.h:16]: As we decouple `DLDevice_` from `XPUHooks`, here we remove it.

File: aten/src/ATen/detail/XPUHooksInterface.h [hunk 13]
```diff
@@ -44,23 +39,8 @@ struct TORCH_API XPUHooksInterface {
...
-  virtual Device getATenDeviceFromDLPackDevice(
-      const DLDevice_& dl_device,
-      void* data) const {
-    TORCH_CHECK(
-        false,
-        "Cannot get XPU device without Intel Extension for Pytorch. ",
-        XPU_HELP);
-  }
-
-  virtual DLDevice_& getDLPackDeviceFromATenDevice(
```
Comment (guangyey) [inline aten/src/ATen/detail/XPUHooksInterface.h:56]: We replaced `getDLPackDeviceFromATenDevice` with `getGlobalIdxFromDevice`

File: aten/src/ATen/detail/XPUHooksInterface.h [hunk 14]
```diff
@@ -44,23 +39,8 @@ struct TORCH_API XPUHooksInterface {
         XPU_HELP);
   }
 
-  virtual Device getATenDeviceFromDLPackDevice(
```
Comment (guangyey) [inline aten/src/ATen/detail/XPUHooksInterface.h:47]: We replace `getATenDeviceFromDLPackDevice` with `getDeviceFromPtr`

File: aten/src/ATen/dlpack.h [hunk 15]
```diff
@@ -94,10 +94,7 @@ typedef enum {
 /*!
  * \brief A Device for Tensor and operator.
  */
-// NB: This is the only difference from
-// https://github.com/dmlc/dlpack/blob/v0.7/include/dlpack/dlpack.h Required to
-// allow forward declaration of DLDevice.
-typedef struct DLDevice_ {
```
Comment (guangyey) [inline aten/src/ATen/dlpack.h:100]: No need for forward declaration for `XPUHooksInterface` anymore, revert `DLDevice_` to its original state.

**Review by EikanWang — APPROVED** [review]

File: aten/src/ATen/DLConvertor.cpp [hunk 16]
```diff
@@ -129,7 +130,7 @@ static Device getATenDevice(const DLDevice& ctx, void* data) {
       return at::Device(DeviceType::HIP, ctx.device_id);
 #endif
     case DLDeviceType::kDLOneAPI:
-      return at::detail::getXPUHooks().getATenDeviceFromDLPackDevice(ctx, data);
+      return at::detail::getXPUHooks().getDeviceFromPtr(data);
```
Comment (jgong5) [inline aten/src/ATen/DLConvertor.cpp:133]: Can you explain why we cannot do `at::Device(DeviceType::XPU, ctx.device_id)` and has to be different from other devices here?

Comment (guangyey) [reply aten/src/ATen/DLConvertor.cpp:133]: The main reason is that the SYCL specification does not explicitly state that the order of enumeration for SYCL devices is fixed. The external program may not necessarily correspond to the same order as device index in PyTorch. So we have to be cautious about this and find the target device via data pointer.
Please correct me if I am wrong.@gujinghui

**Comment by huydhn** [comment]
@pytorch rebase

**Comment by huydhn** [comment]
You can comment `@pytorchbot rebase` to rebase the PR to main.

**Review by jgong5 — APPROVED** [review]

File: aten/src/ATen/xpu/detail/XPUHooks.cpp [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+bool XPUHooks::hasXPU() const {
+  return true;
+}
+
+std::string XPUHooks::showConfig() const {
+  return "XPU backend";
+}
+
+DeviceIndex XPUHooks::getGlobalIdxFromDevice(const at::Device& device) const {
```
Comment (gujinghui) [inline aten/src/ATen/xpu/detail/XPUHooks.cpp:20]: Should we use int as return type? The global index is not DeviceIndex defined by PyTorch, right?
Instead, it's defined by DLPack? We cannot guarantee it always aligns to PyTorch definition?

Comment (EikanWang) [reply aten/src/ATen/xpu/detail/XPUHooks.cpp:20]: @guangyey , as we discussed, we should return int and add some comments here, right?

Comment (guangyey) [reply aten/src/ATen/xpu/detail/XPUHooks.cpp:20]: Agreed. change to return int.

File: aten/src/ATen/xpu/XPUContext.cpp [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+ * Device properties are lazily initialized when the first time properties are
+ * requested for a device.
+ */
+DeviceIndex num_gpus = -1;
+c10::once_flag init_flag;
+std::deque<c10::once_flag> device_prop_flags;
+std::vector<DeviceProp> device_properties;
+
+std::deque<c10::once_flag> device_global_idx_flags;
+std::vector<int> device_global_idxs;
```
Comment (guangyey) [inline aten/src/ATen/xpu/XPUContext.cpp:26]: `device_global_idxs` is used in DLPack, so use `int` as its data type.

**Comment by EikanWang** [comment]
> You can comment `@pytorchbot rebase` to rebase the PR to main.

Thanks. Yes, we will rebase the PR as soon as the first PR(#116019 ) being landed

**Comment by EikanWang** [comment]
@malfet , may I know if you have any comments on this ghstack?

**Review by EikanWang — APPROVED** [review]

**Review by gujinghui — APPROVED** [review]

**Comment by chuanqi129** [comment]
Hi @atalman @huydhn, seems the xpu test workflow has permission issue https://github.com/pytorch/pytorch/actions/runs/7502988971, could you please help to double check it?
```
Invalid workflow file: .github/workflows/xpu.yml#L29
The workflow is not valid. .github/workflows/xpu.yml (Line: 29, Col: 3): Error calling workflow 'pytorch/pytorch/.github/workflows/_xpu-test.yml@8a32701c9d734128e7f891615ff956d3fa1406c0'. The workflow is requesting 'id-token: write', but is only allowed 'id-token: none'.
```

File: aten/src/ATen/Context.h [hunk 19]
```diff
@@ -477,19 +479,6 @@ static inline void manual_seed(uint64_t seed) {
     }
   }
 
-  const auto xpu_num_gpus = detail::getXPUHooks().getNumGPUs();
```
Comment (guangyey) [inline aten/src/ATen/Context.h:480]: These codes will be added with PR related to `Generator` as they affect `torch.manual_seed` in CI.

**Comment by malfet** [comment]
Changed the topic, as "Not user facing" feels like a wrong topic for PRs that attempts to introduce new features, ins't it?

**Comment by EikanWang** [comment]
> Changed the topic, as "Not user facing" feels like a wrong topic for PRs that attempts to introduce new features, ins't it?

Oh, yes. You mean users need to be aware of `torch.xpu`, right?

File: aten/src/ATen/xpu/XPUContext.cpp (hunk 18 above)
Comment (malfet) [reply aten/src/ATen/xpu/XPUContext.cpp:26]: Would be good to be a bit more explicit, for example use `int32_t`

File: aten/src/ATen/xpu/XPUContext.cpp [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+  device_prop_flags.resize(num_gpus);
+  device_properties.resize(num_gpus);
+  device_global_idx_flags.resize(num_gpus);
+  device_global_idxs.resize(num_gpus);
+}
+
+void initDeviceProperty(DeviceIndex device) {
+  DeviceProp device_prop;
+  c10::xpu::get_device_properties(&device_prop, device);
+  device_properties[device] = device_prop;
```
Comment (malfet) [inline aten/src/ATen/xpu/XPUContext.cpp:39]: Why not do that?
```suggestion
  c10::xpu::get_device_properties(&device_properties[device], device);
```

File: aten/src/ATen/xpu/XPUContext.cpp [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+      static_cast<int>(std::distance(devices.begin(), it));
+}
+
+inline void check_device(DeviceIndex device) {
+  TORCH_CHECK(
+      device >= 0 && device < num_gpus,
+      "device is out of range, device is ",
+      static_cast<int>(device),
+      ", total number of device is ",
+      static_cast<int>(num_gpus),
```
Comment (malfet) [inline aten/src/ATen/xpu/XPUContext.cpp:62]: Why do you need to cast here?
```suggestion
      num_gpus,
```

**Review by malfet — APPROVED** [review]
Looks OK to me, though it feels it mostly just copy-n-pastes code from `aten/src/ATen/cuda/` and renames `CUDA` to `XPU` 
If this is the case, would it be possible to use templates of sort to avoid code duplication?

File: aten/src/ATen/xpu/XPUContext.cpp (hunk 21 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUContext.cpp:62]: `device` and `num_gpus` are `DeviceIndex` types, which is `int8_t`. `int8_t` is similar as `signed char` and `std::cout` can not correctly print its value. refer to https://stackoverflow.com/questions/7587782/why-is-stdcout-not-printing-the-correct-value-for-my-int8-t-number
According to the existing PyTorch [code](https://github.com/pytorch/pytorch/blob/1872834247084d76544c95e04fe2ec3370bac87d/c10/core/DeviceType.cpp#L58) in `c10/core/DeviceType.cpp`, I will change `static_cast<int>(num_gpus)` to `static_cast<int16_t>(num_gpus)`.

File: aten/src/ATen/xpu/XPUContext.cpp (hunk 20 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUContext.cpp]: Thanks, done.
